### PR TITLE
Log if request reaches the end of request pipeline

### PR DESCRIPTION
**Pull request by JamesNK**
Contributes to https://github.com/dotnet/aspnetcore/issues/46404

### CODE REVIEW

File: src/Http/Http/src/Builder/ApplicationBuilder.cs [hunk 1]
```diff
@@ -149,4 +165,12 @@ public RequestDelegate Build()
 
         return app;
     }
+
+    private static partial class Log
+    {
+        [LoggerMessage(1, LogLevel.Information,
+            "Request reached the end of the middleware pipeline {Protocol} {Method} {Scheme}://{Host}{PathBase}{Path}",
```
Comment (BrennanConroy) [inline src/Http/Http/src/Builder/ApplicationBuilder.cs:172]: This sounds too generic.
Suggestion:
`"Request reached the end of the middleware pipeline without being handled by application code. Returning status {StatusCode}. {Protocol} {Method} etc..."`

**Comment by JamesNK** [comment]
@Tratcher

File: src/Http/Http/src/Builder/ApplicationBuilder.cs [hunk 2]
```diff
@@ -149,4 +167,12 @@ public RequestDelegate Build()
 
         return app;
     }
+
+    private static partial class Log
+    {
+        [LoggerMessage(1, LogLevel.Information,
+            "Request reached the end of the middleware pipeline without being handled by application code {Protocol} {Method} {Scheme}://{Host}{PathBase}{Path} - {statusCode}",
```
Comment (BrennanConroy) [inline src/Http/Http/src/Builder/ApplicationBuilder.cs:174]:
```suggestion
            "Request reached the end of the middleware pipeline without being handled by application code {Protocol} {Method} {Scheme}://{Host}{PathBase}{Path} - {StatusCode}",
```

File: src/Http/Http/test/ApplicationBuilderTests.cs [hunk 3]
```diff
@@ -74,6 +76,55 @@ public async Task BuildImplicitlyThrowsForMatchedEndpointAsLastStep()
...
+        await app.Invoke(httpContext);
+
+        Assert.Equal(404, httpContext.Response.StatusCode);
+
+        var log = TestSink.Writes.Single(w => w.EventId.Name == "RequestPipelineEnd");
+        Assert.Equal("Request reached the end of the middleware pipeline without being handled by application code HTTP/2 GET https://localhost:5000/pathbase/path - 404", log.Message);
+    }
+
+    [Fact]
+    public async Task BuildDontLogOrChangeStatusWithTerminalMiddleware()
```
Comment (BrennanConroy) [inline src/Http/Http/test/ApplicationBuilderTests.cs:107]: nit:
```suggestion
    public async Task BuildDoesNotLogOrChangeStatusWithTerminalMiddleware()
```

**Review by BrennanConroy — APPROVED** [review]

File: src/Http/Http/src/Builder/ApplicationBuilder.cs [hunk 4]
```diff
@@ -149,4 +167,12 @@ public RequestDelegate Build()
 
         return app;
     }
+
+    private static partial class Log
+    {
+        [LoggerMessage(1, LogLevel.Information,
+            "Request reached the end of the middleware pipeline without being handled by application code {Protocol} {Method} {Scheme}://{Host}{PathBase}{Path} - {StatusCode}",
```
Comment (JamesNK) [inline src/Http/Http/src/Builder/ApplicationBuilder.cs:174]: @Tratcher What do you think of adding query string to the log? Everything else is here. I think people will find it confusing if the query string is unexplainable missing.

Comment (mitchdenny) [reply src/Http/Http/src/Builder/ApplicationBuilder.cs:174]: Could that be a security issue? Some people put secrets on query strings (e.g. SAS tokens). If someone has a SAS token in their query string, but the base path is incorrect then you could start leaking those active credentials into your logs.

Comment (JamesNK) [reply src/Http/Http/src/Builder/ApplicationBuilder.cs:174]: There are existing logs that include the query string: https://github.com/dotnet/aspnetcore/blob/e53379914e7ed68eab66f5dac5c8682ac6f3545e/src/Hosting/Hosting/src/Internal/HostingRequestStartingLog.cs#L12

Comment (Tratcher) [reply src/Http/Http/src/Builder/ApplicationBuilder.cs:174]: Don't. Queries tend to be large, less impactful on routing, and to contain more PII.

Comment (Tratcher) [reply src/Http/Http/src/Builder/ApplicationBuilder.cs:174]: Protocol isn't relevant to routing either is it?

**Review by Tratcher — APPROVED** [review]

**Pull request merged**